syntax = "proto3";

package mlos.optimizer_monitoring_service;

import public "OptimizerShared.proto";

// Exposes Bayesian Optimizer's functionality over gRPC.
//
service OptimizerMonitoringService
{
    // Returns a list of existing optimizer instances.
    //
    rpc ListExistingOptimizers(Empty) returns (OptimizerList);

    // Returns information about optimizers configuration and optimization problem.
    //
    rpc GetOptimizerInfo(OptimizerHandle) returns (OptimizerInfo);

    // Returns the current optimizer convergence state.
    //
    rpc GetOptimizerConvergenceState(OptimizerHandle) returns (OptimizerConvergenceState);

    // Returns the computed goodness of fit metrics.
    //
    rpc ComputeGoodnessOfFitMetrics(OptimizerHandle) returns (SimpleString);

    // Check if optimizer has been trained.
    //
    rpc IsTrained(OptimizerHandle) returns (SimpleBoolean);

    // Adds observations to the optimizer's data set.
    //
    rpc RegisterObservations(RegisterObservationsRequest) returns (Empty);

    // Produces a prediction for specified features.
    //
    rpc Predict(PredictRequest) returns (PredictResponse);

    // Returns all observations registered for a given optimizer.
    //
    rpc GetAllObservations(OptimizerHandle) returns (Observations);

    // Like ping.
    //
    rpc Echo(Empty) returns (Empty);
};

// Represents information about optimizer convergence state.
//
message OptimizerConvergenceState
{
    OptimizerHandle OptimizerHandle = 1;
    string SerializedOptimizerConvergenceState = 2;
};

// A list of OptimizerInfo messages.
//
message OptimizerList
{
    repeated OptimizerInfo Optimizers = 1;
}


// A request to predict objective values given specified features (parameters and context).
//
message PredictRequest
{
    OptimizerHandle OptimizerHandle = 1;
    Features Features = 2;
};

// Representation of an objective value prediction
//
message SingleObjectivePrediction
{
    string ObjectiveName = 1;
    string PredictionDataFrameJsonString = 2;
};

message PredictResponse
{
    // repeated for every objective (in a multi-objective case)
    //
    repeated SingleObjectivePrediction ObjectivePredictions = 1;
};


